ProcessLib/Process.h

```diff
@@ -12,35 +12,180 @@
 
 #include <string>
 
+#include <boost/optional.hpp>
+
+#include "AssemblerLib/ComputeSparsityPattern.h"
+#include "AssemblerLib/VectorMatrixAssembler.h"
+#include "AssemblerLib/LocalToGlobalIndexMap.h"
+#include "BaseLib/ConfigTree.h"
+#include "MathLib/LinAlg/SetMatrixSparsity.h"
+#include "MeshLib/MeshSubsets.h"
+
+#ifdef USE_PETSC
+#include "MeshLib/NodePartitionedMesh.h"
+#include "MathLib/LinAlg/PETSc/PETScMatrixOption.h"
+#endif
+
+#include "ProcessVariable.h"
+
 namespace MeshLib
 {
-    class Mesh;
+class Mesh;
 }
 
 namespace ProcessLib
 {
-
+template <typename GlobalSetup>
 class Process
 {
 public:
-    Process(MeshLib::Mesh& mesh)
-        : _mesh(mesh)
-    { }
+	Process(MeshLib::Mesh& mesh) : _mesh(mesh) {}
+	virtual ~Process()
+	{
+		for (auto p : _all_mesh_subsets)
+			delete p;
+	}
+
+	/// Process specific initialization called by initialize().
+	virtual void init() = 0;
+	virtual bool assemble(const double delta_t) = 0;
+
+	virtual std::string getLinearSolverName() const = 0;
+
+	/// Postprocessing after solve().
+	/// The file_name is indicating the name of possible output file.
+	virtual void post(std::string const& file_name) = 0;
+	virtual void postTimestep(std::string const& file_name,
+	                          const unsigned timestep) = 0;
+
+	/// Creates mesh subsets, i.e. components, for given mesh.
+	virtual void initializeMeshSubsets(MeshLib::Mesh const& mesh) = 0;
+
+	void initialize()
+	{
+		DBUG("Initialize process.");
+
+		DBUG("Construct dof mappings.");
+		initializeMeshSubsets(_mesh);
+
+		_local_to_global_index_map.reset(
+		    new AssemblerLib::LocalToGlobalIndexMap(
+		        _all_mesh_subsets, AssemblerLib::ComponentOrder::BY_COMPONENT));
+
+#ifndef USE_PETSC
+		DBUG("Compute sparsity pattern");
+		computeSparsityPattern();
+#endif
+
+		// create global vectors and linear solver
+		createLinearSolver(getLinearSolverName());
+
+		DBUG("Create global assembler.");
+		_global_assembler.reset(
+		    new GlobalAssembler(*_A, *_rhs, *_local_to_global_index_map));
 
-    virtual ~Process() = default;
+		init();  // Execute proces specific initialization.
+	}
 
-    virtual void initialize() = 0;
-    virtual bool solve(const double delta_t) = 0;
+	void initializeNeumannBcs(std::vector<NeumannBc<GlobalSetup>*> const& bcs)
+	{
+		for (auto bc : bcs)
+			bc->initialize(_global_setup, *_A, *_rhs, _mesh.getDimension());
+	}
 
-    /// Postprocessing after solve().
-    /// The file_name is indicating the name of possible output file.
-    virtual void post(std::string const& file_name) = 0;
-    virtual void postTimestep(std::string const& file_name, const unsigned timestep) = 0;
+	bool solve(const double delta_t)
+	{
+		_A->setZero();
+		MathLib::setMatrixSparsity(*_A, _sparsity_pattern);
+
+		bool const result = assemble(delta_t);
+
+		_linear_solver->solve(*_rhs, *_x);
+		return result;
```

Comment:
This is in horrible state now, I know. The reason for the return value is to report to nonlinear solver or timestepper that current damping or delta-t are too large. Will fix later. Todo added.